File: packages/distance/src/edit/sift4.ts
/**
 * SIFT4 simple — fast approximate string distance.
 *
 * A fast algorithm for approximate string matching with O(n) complexity
 * in typical cases. Returns a distance value (lower = more similar).
 *
 * Matches the textdistance.rs implementation exactly.
 *
 * Time: O(n * maxOffset)
 */

import { normalize } from "../utils";

/**
 * Options for SIFT4.
 */
export interface ISift4Options {
  /**
   * Maximum offset for character matching.
   * @default 5
   */
  maxOffset?: number;
}

/**
 * Compute the SIFT4 simple distance between two strings.
 *
 * @param a - First string
 * @param b - Second string
 * @param options - Configuration
 * @returns Distance (non-negative integer)
 */
export function sift4(a: string, b: string, options: ISift4Options = {}): number {
  const maxOffset = options.maxOffset ?? 5;

  const aLen = a.length;
  const bLen = b.length;

  let c1 = 0;
  let c2 = 0;
  let lcss = 0;
  let localCs = 0;

  while (c1 < aLen && c2 < bLen) {
    if (a.charCodeAt(c1) === b.charCodeAt(c2)) {
      localCs++;
    } else {
      lcss += localCs;
      localCs = 0;
      if (c1 !== c2) {
        c1 = Math.min(c1, c2);
        c2 = c1;
      }

      for (let offset = 0; offset < maxOffset; offset++) {
        if (!(c1 + 1 < aLen || c2 + offset < bLen)) break;

        if (c1 + offset < aLen && a.charCodeAt(c1 + offset) === b.charCodeAt(c2)) {
          c1 += offset;
          localCs++;
          break;
        }
        if (c2 + offset < bLen && a.charCodeAt(c1) === b.charCodeAt(c2 + offset)) {
          c2 += offset;
          localCs++;
          break;
        }
      }
    }
    c1++;
    c2++;
  }

  return Math.max(aLen, bLen) - lcss - localCs;
}

/**
 * Compute the normalized SIFT4 similarity in [0, 1].
 *
 * @param a - First string
 * @param b - Second string
 * @param options - Configuration
 * @returns Similarity score where 1 means identical
 */
export function sift4Normalized(a: string, b: string, options: ISift4Options = {}): number {
  return normalize(sift4(a, b, options), Math.max(a.length, b.length));
}

File: packages/distance-wasm/src/edit/sift4.rs
/// SIFT4 simple — fast approximate string distance.
///
/// Time: O(n * maxOffset)

use crate::utils::normalize;

const MAX_OFFSET: usize = 5;

pub fn sift4_simple(a: &str, b: &str) -> u32 {
    let a_bytes = a.as_bytes();
    let b_bytes = b.as_bytes();
    let a_len = a_bytes.len();
    let b_len = b_bytes.len();

    let mut c1 = 0usize;
    let mut c2 = 0usize;
    let mut lcss = 0u32;
    let mut local_cs = 0u32;

    while c1 < a_len && c2 < b_len {
        if a_bytes[c1] == b_bytes[c2] {
            local_cs += 1;
        } else {
            lcss += local_cs;
            local_cs = 0;
            if c1 != c2 {
                c1 = c1.max(c2);
                c2 = c1;
                if c1 >= a_len || c2 >= b_len { break; }
            }
            for offset in 0..MAX_OFFSET {
                if c1 + offset >= a_len && c2 + offset >= b_len { break; }
                if c1 + offset < a_len && a_bytes[c1 + offset] == b_bytes[c2] {
                    c1 += offset;
                    local_cs += 1;
                    break;
                }
                if c2 + offset < b_len && a_bytes[c1] == b_bytes[c2 + offset] {
                    c2 += offset;
                    local_cs += 1;
                    break;
                }
            }
        }
        c1 += 1;
        c2 += 1;
    }

    lcss += local_cs;
    (a_len.max(b_len) as u32) - lcss
}

pub fn sift4_simple_normalized(a: &str, b: &str) -> f64 {
    let max_len = a.len().max(b.len()) as u32;
    normalize(sift4_simple(a, b), max_len)
}
